File: app/securesocial/provider/providers/LinkedInProvider.java
/**
 * Copyright 2011 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.provider.providers;

import com.google.gson.JsonObject;
import com.google.gson.JsonElement;
import play.Logger;
import play.libs.WS;
import securesocial.provider.*;

import java.util.Map;

/**
 * A provider for LinkedIn
 */
public class LinkedInProvider extends OAuth1Provider
{
    private static final String ME_API = "https://api.linkedin.com/v1/people/~:(id,first-name,last-name,picture-url)?format=json";
    private static final String ERROR_CODE = "errorCode";
    private static final String MESSAGE = "message";
    private static final String REQUEST_ID = "requestId";
    private static final String TIMESTAMP = "timestamp";
    private static final String ID = "id";
    private static final String FIRST_NAME = "firstName";
    private static final String LAST_NAME = "lastName";
    private static final String PICTURE_URL = "pictureUrl";

    public LinkedInProvider() {
        super(ProviderType.linkedin);
    }

    @Override
    protected void fillProfile(SocialUser user, Map<String, Object> authContext) {
        JsonObject me = WS.url(ME_API).oauth(user.serviceInfo,user.token, user.secret).get().getJson().getAsJsonObject();

        if ( me.get(ERROR_CODE) != null ) {
            int errorCode = me.get(ERROR_CODE).getAsInt();
            final String message = me.get(MESSAGE).getAsString();
            final String requestId = me.get(REQUEST_ID).getAsString();
            final String timestamp = me.get(TIMESTAMP).getAsString();
            Logger.error("Error retrieving profile information from LinkedIn. Error code: %s, message: %s, requestId: %s, timestamp: %s.",
                    errorCode, message, requestId, timestamp);
            throw new AuthenticationException();
        }
        user.id.id = me.get(ID).getAsString();
        user.displayName = FoursquareProvider.fullName(me.get(FIRST_NAME).getAsString(),me.get(LAST_NAME).getAsString());
        JsonElement picture = me.get(PICTURE_URL);
        if(picture != null) {
            user.avatarUrl = picture.getAsString();
        }
        // can't get the email because the LinkedIn API does not provide it.
        //user.email = ;
    }
}

File: project-code/app/securesocial/core/providers/LinkedInProvider.scala
/**
 * Copyright 2012 Jorge Aliss (jaliss at gmail dot com) - twitter: @jaliss
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package securesocial.core.providers

import securesocial.core._
import play.api.mvc.{Request, Results, Result}
import play.api.libs.oauth.{RequestToken, OAuthCalculator}
import play.api.libs.ws.{Response, WS}
import play.api.{Application, Logger}


/**
 * A LinkedIn Provider
 */
class LinkedInProvider(application: Application) extends OAuth1Provider(application) {


  override def providerId = LinkedInProvider.LinkedIn

  override def fillProfile(user: SocialUser): SocialUser = {
    val oauthInfo = user.oAuth1Info.get
    WS.url(LinkedInProvider.Api).sign(OAuthCalculator(oauthInfo.serviceInfo.key,
      RequestToken(oauthInfo.token, oauthInfo.secret))).get().await(10000).fold(
      onError => {
        Logger.error("timed out waiting for LinkedIn")
        throw new AuthenticationException()
      },
      response =>
      {
        val me = response.json
        (me \ "errorCode").asOpt[Int] match {
          case Some(error) => {
            val message = (me \ "message").asOpt[String]
            val requestId = (me \ "requestId").asOpt[String]
            val timestamp = (me \ "timestamp").asOpt[String]
            Logger.error(
              "Error retrieving information from LinkedIn. Error code: %s, requestId: %s, message: %s, timestamp: %s"
              format(error, message, requestId, timestamp)
            )
            throw new AuthenticationException()
          }
          case _ => {
            val id = (me \ "id").as[String]
            val first = (me \ "firstName").asOpt[String]
            val last = (me \ "lastName").asOpt[String]
            val fullName = "%s %s".format(first.getOrElse(""), last.getOrElse(""))
            val avatarUrl = (me \ "pictureUrl").asOpt[String]
            user.copy(id = UserId(id, providerId), displayName = fullName, avatarUrl = avatarUrl)
          }
        }
      }
    )
  }
}

object LinkedInProvider {
  val Api = "https://api.linkedin.com/v1/people/~:(id,first-name,last-name,picture-url)?format=json"
  val LinkedIn = "linkedin"
}
